text/0113-secure-script-container.md

```diff
@@ -0,0 +1,207 @@
+
+Other considerations for Phase II:
+
+ * the virtual script web component "should" be registered before any shadow root is created, it is not longer a "must", the flip side of this is that pre-existing shadowRoots are just not going to be inspected by virtual scripts.
+ * budget is less than 10kb (minify/gzip).
+ * performance penalties will be added to the virtual scripts as they run inside a membrane:
+   * evaluation budget: 30% slower than evaluating the same script in the main window.
+   * operation budget: 1X slower than executing the same task is the main window.
+   * 3rd party scripts are rarely part of the critical path for the application to function.
+ * synthetic shadow and native shadow combinations can be allowed, and supported, although it will require an explicit order to run before synthetic shadow polyfill does it.
+
+### Globals
+
+Some scripts define globals which are meant to be used by the rest of the application. Let's take this example of Google Analytics:
+
+```html
+<!-- Google Analytics -->
+<script>
+    window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
+    ga('create', 'UA-XXXXX-Y', 'auto');
+    ga('send', 'pageview');
+</script>
+<script async src='https://www.google-analytics.com/analytics.js'></script>
+<!-- End Google Analytics -->
+```
+
+If we evaluate the above snippets with `<virtual-script>` instead, then `ga` global value will only be available into the sandbox. So `ga` will not be usable by anyone else in the entire application. We can however, allow for globals to be exposed into the main window. A syntax like the following could be provided:
+
+```html
+<virtual-script extra-globals="ga"></virtual-script>
+```
+
+- `virtual-script` container would not have access to globals defined by other scripts in the outer window unless they were defined in a `virtual-script` using `extra-globals`.
+- Globals defined in `extra-globals` will be redefined in any `virtual-script` containers.
+- They will also be available to any Locker-created sandboxes
+
+Note: these global values are controlled side-channels between the main window and the sandbox, and should be used with caution. For example, creating a component that depends on `ga` to be defined, is an anti-pattern.
+
+### Supported Attributes
+
+`crossorigin`: indicates that the browser should provide more information to window.onerror when handling CORS errors
+
+`defer`: indicates that the script should be executed after the document has parsed (has no effect if `src` is absent)
+
+`integrity`: enables the browser to verify thata fetched resource is delivered without unexpected manipulation
+
+`nomodule`: the script should not be executed in browsers that support ES2015 modules
+
+`nonce`: user provided nonce values for CSP will be supported
```

Comment:
Will we be able to support CSP's `<hash-algorithm>-<base64-value>` to enable execution of inline virtual scripts?

Reply:
As long as virtual-script doesn't mutate the textContent of the tag then it shouldn't have an impact on that.